src/core/ledger.rs

```diff
@@ -69,91 +70,63 @@ pub struct Ledger {
 
     pub(crate) directives: Vec<Spanned<Directive>>,
     pub metas: Vec<Spanned<Directive>>,
+
+    transformer: Arc<dyn Transformer>,
 }
 
 impl Ledger {
-    pub async fn load(entry: PathBuf, endpoint: String) -> ZhangResult<Ledger> {
-
-        Ledger::load_with_database(entry, endpoint, None).await
+    pub async fn load<T: Transformer + Default + 'static>(entry: PathBuf, endpoint: String) -> ZhangResult<Ledger> {
+        Ledger::load_with_database::<T>(entry, endpoint, None).await
     }
 
-    pub async fn load_with_database(entry: PathBuf, endpoint: String, database: Option<PathBuf>) -> ZhangResult<Ledger> {
-        let entry = entry.canonicalize().with_path(&entry)?;
-        let main_endpoint = entry.join(&endpoint);
-        let main_endpoint = main_endpoint.canonicalize().with_path(&main_endpoint)?;
-        let mut load_queue = VecDeque::new();
-        load_queue.push_back(main_endpoint);
-
-        let mut visited = HashSet::new();
-        let mut directives = vec![];
-        while let Some(load_entity) = load_queue.pop_front() {
-            let path = load_entity.canonicalize().with_path(&load_entity)?;
-            debug!("visited entry file: {}", path.to_str().unwrap());
-            if visited.contains(&path) {
-                continue;
-            }
-            let entity_directives = Ledger::load_directive_from_file(load_entity)?;
-            entity_directives
-                .iter()
-                .filter(|it| it.directive_type() == DirectiveType::Include)
-                .for_each(|it| match &it.data {
-                    Directive::Include(include_directive) => {
-                        let buf = PathBuf::from(include_directive.file.clone().to_plain_string());
-                        let include_path = path.parent().map(|it| it.join(&buf)).unwrap_or(buf);
-                        load_queue.push_back(include_path)
-                    }
-                    _ => {
-                        unreachable!()
-                    }
-                });
-            visited.insert(path);
-            directives.extend(entity_directives)
-        }
+    pub async fn load_with_database<T: Transformer + Default + 'static>(
```

Comment:
My advice is to form a tree of directives in these `load` functions (i.e. each node is a file and its corresponding directives, and branches are generated when encountering `Include` directives by using DFS), then feed the tree to the actual transformer, and finally flatten the tree to a list. Transformers can avoid file operations fully and focus on dealing with directives in this way.

Reply:
As the original design, transformer is not for text source specifically, it represent how the source data be tranformed to zhang's `Directive`. 
I get your idea's props, which is to separate file operation and the data transform operation. 
I will say we can introduce a new trait `FileTransformer` based on the `Transformer` to handle file operations.

```rust

trait Tranformer {
  type Source;  // Source can be any type
  fn load(&self, entry: Self::Source, endpoint: String) -> ZhangResult<TransformResult>
}

trait FileTranfomer: Transformer {
type Source = PathBuf;
  fn load(&self, entry: PathBuf, endpoint: String) -> ZhangResult<TransformResult> {
    // default implementation, read entry files to a tree in DFS manner
    let tree: FileTreeNode = ...;
    Ok(TransformResult(self.transform(tree), visited_files)) // error handling omitted
  }

  // leave for subtypes
  fn transform(&self, tree: &FileTreeNode) -> ZhangResult<Vec<Directives>>;
}
struct ZhangTextTransformer;

impl FileTranfomer for ZhangTextTransformer {
  fn transform(&self, tree: &FileTreeNode) -> ZhangResult<Vec<Directives>> {
    // just flatten the tree with some additional checks (e.g. report error on unsupported directives)
  }
}

struct BeancountTextTransformer;

impl FileTranfomer for BeancountTextTransformer {
  fn transform(&self, tree: &FileTreeNode) -> ZhangResult<Vec<Directives>> {
    // traverse the tree, convert beancount directives (e.g. merge balance with following pad, use a context to store current tags) 
  }
}

Reply:
This extra layer of abstraction seems to be good! The name could be something like `TreeTextFileTransformer` to be cleaner.